In [1]:
import pandas as pd 
import numpy as np 
import seaborn as sn


In [2]:
# What's the version of Pandas that you installed?

pd.__version__

'2.0.3'

In [4]:
# Load the dataset

df = pd.read_csv('https://raw.githubusercontent.com/alexeygrigorev/datasets/master/housing.csv')
df.head()

longitude  latitude  housing_median_age  total_rooms  total_bedrooms  \
0    -122.23     37.88                41.0        880.0           129.0   
1    -122.22     37.86                21.0       7099.0          1106.0   
2    -122.24     37.85                52.0       1467.0           190.0   
3    -122.25     37.85                52.0       1274.0           235.0   
4    -122.25     37.85                52.0       1627.0           280.0   

   population  households  median_income  median_house_value ocean_proximity  
0       322.0       126.0         8.3252            452600.0        NEAR BAY  
1      2401.0      1138.0         8.3014            358500.0        NEAR BAY  
2       496.0       177.0         7.2574            352100.0        NEAR BAY  
3       558.0       219.0         5.6431            341300.0        NEAR BAY  
4       565.0       259.0         3.8462            342200.0        NEAR BAY

In [8]:
# How many columns are in the dataset?

print(f"The dataset has {df.shape[1]} columns")

The dataset has 10 columns


In [12]:
# Which columns in the dataset have missing values 

df.isnull().sum().sort_values(ascending=False)

total_bedrooms        207
longitude               0
latitude                0
housing_median_age      0
total_rooms             0
population              0
households              0
median_income           0
median_house_value      0
ocean_proximity         0
dtype: int64

you can see that total_bedrooms has 207 missing value

In [16]:
# How many unique values does the ocean_proximity column have 

print(f"ocean_proximity has {df['ocean_proximity'].nunique()} unique values")

ocean_proximity has 5 unique values


In [24]:
# What's the average value of the median_house_value 
# for the houses located near the bay

house_near_bay = df[df['ocean_proximity'] == 'NEAR BAY']
avg_median_house_value = house_near_bay['median_house_value'].mean()
print(f'the average value of the median_house_value for the houses located near the bay is {avg_median_house_value:.2f}')


the average value of the median_house_value for the houses located near the bay is 259212.31


1. Calculate the average of total_bedrooms column in the dataset.
2. Use the fillna method to fill the missing values in total_bedrooms with the mean value from the previous step.
3. Now, calculate the average of total_bedrooms again.
4. Has it changed?

In [26]:
# avg total_bedrooms

avg_total_bedrooms = df['total_bedrooms'].mean()

# handling missing value 

df2 = df.copy()
df2['total_bedrooms'].fillna(avg_total_bedrooms, axis = 0)

# avg total_bedrooms in df2
avg_total_bedrooms_2 = df2['total_bedrooms'].mean()

print(f'average total_bedrooms before the missing value handled is {avg_total_bedrooms:.3f}')
print(f'average total_bedrooms after the missing value handled is {avg_total_bedrooms_2:.3f}')


average total_bedrooms before the missing value handled is 537.871
average total_bedrooms after the missing value handled is 537.871


1. Select all the options located on islands.
2. Select only columns housing_median_age, total_rooms, total_bedrooms.
3. Get the underlying NumPy array. Let's call it X.
4. Compute matrix-matrix multiplication between the transpose of X and X. To get the transpose, use X.T. Let's call the result XTX.
5. Compute the inverse of XTX.
6. Create an array y with values [950, 1300, 800, 1000, 1300].
7. Multiply the inverse of XTX with the transpose of X, and  then multiply the result by y. Call the result w.
8. What's the value of the last element of w?

In [28]:
house_on_island = df2[df2['ocean_proximity'] == 'ISLAND'][['housing_median_age', 'total_rooms', 'total_bedrooms']]
X = house_on_island.to_numpy()
XTX = np.dot(X.T, X)
XTX_inverse = np.linalg.inv(XTX)
y = np.array([950, 1300, 800, 1000, 1300])
w = np.dot(np.dot(XTX_inverse, X.T), y)
last_element_of_w = w[-1]
print(f'Value of the last element of w: {last_element_of_w}')



Value of the last element of w: 5.6992294550655656
